components/feature/sendtab/src/main/java/mozilla/components/feature/sendtab/SendTabUseCases.kt

```diff
@@ -0,0 +1,109 @@
+import mozilla.components.concept.sync.TabData
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.support.ktx.kotlin.crossProduct
+
+/**
+ * Contains use cases for sending tabs to devices related to the firefox-accounts.
+ */
+class SendTabUseCases(accountManager: FxaAccountManager) {
+
+    class SendToDeviceUseCase internal constructor(
+        private val accountManager: FxaAccountManager
+    ) {
+        /**
+         * Sends the tab to provided deviceId if possible.
+         *
+         * @param deviceId The deviceId to send the tab.
+         * @param tab The tab to send.
+         */
+        operator fun invoke(deviceId: String, tab: TabData) {
+            filterSendTabDevices(accountManager) { constellation, devices ->
+                val device = devices.firstOrNull {
+                    it.id == deviceId
+                }
+                device?.let {
+                    constellation.sendEventToDeviceAsync(device.id, SendTab(tab.title, tab.url))
+                }
+            }
+        }
+
+        /**
+         * Sends the tabs to provided deviceId if possible.
+         *
+         * @param deviceId The deviceId to send the tab.
+         * @param tabs The list of tabs to send.
+         */
+        operator fun invoke(deviceId: String, tabs: List<TabData>) {
+            filterSendTabDevices(accountManager) { constellation, devices ->
+                val device = devices.firstOrNull { it.id == deviceId }
+
+                device?.let {
+                    tabs.forEach { tab ->
+                        constellation.sendEventToDeviceAsync(device.id, SendTab(tab.title, tab.url))
+                    }
+                }
+            }
+        }
+    }
+
+    class SendToAllUseCase internal constructor(
+        private val accountManager: FxaAccountManager
+    ) {
+
+        /**
+         * Sends the tab to all send-tab compatible devices.
+         *
+         * @param tab The tab to send.
+         */
+        operator fun invoke(tab: TabData) {
+            filterSendTabDevices(accountManager) { constellation, devices ->
+                devices.forEach { device ->
+                    constellation.sendEventToDeviceAsync(device.id, SendTab(tab.title, tab.url))
+                }
+            }
+        }
+
+        /**
+         * Sends the tabs to all the send-tab compatible devices.
+         */
+        operator fun invoke(tabs: Collection<TabData>) {
+            filterSendTabDevices(accountManager) { constellation, devices ->
+                devices.crossProduct(tabs) { device, tab ->
+                    constellation.sendEventToDeviceAsync(device.id, SendTab(tab.title, tab.url))
+                }
+            }
+        }
+    }
+
+    val sendToDevice: SendToDeviceUseCase by lazy { SendToDeviceUseCase(accountManager) }
+    val sendToAll: SendToAllUseCase by lazy { SendToAllUseCase(accountManager) }
+}
+
+@VisibleForTesting
+internal fun filterSendTabDevices(
```

Comment:
Could this function be `inline`?

Reply:
This was inline but coverage doesn't understand inline functions. I'll make it inline regardless and we can force-merge this since we know it is actually green.

Reply:
👍